source/slang/parameter-binding.cpp

```diff
@@ -2813,135 +2885,200 @@ void generateParameterBindings(
-
-            for( auto& varLayout : parameterInfo->varLayouts )
-            {
-                varLayout->findOrAddResourceInfo(LayoutResourceKind::Uniform)->index = uniformOffset.getFiniteValue();
-            }
-        }
-
-        globalScopeStructLayout->fields.Add(firstVarLayout);
-
-        for( auto& varLayout : parameterInfo->varLayouts )
-        {
-            globalScopeStructLayout->mapVarToLayout.Add(varLayout->varDecl.getDecl(), varLayout);
-        }
+        auto entryPointParamsLayout = entryPoint->parametersLayout;
+        completeBindingsForParameter(&context, entryPointParamsLayout);
     }
-    globalScopeRules->EndStructLayout(&structLayoutInfo);
 
-    RefPtr<TypeLayout> globalScopeLayout = globalScopeStructLayout;
-
-    // If there are global-scope uniforms, then we need to wrap
-    // up a global constant buffer type layout to hold them
-    if( needDefaultConstantBuffer )
+    // Next we need to create a type layout to reflect the information
+    // we have collected, and we will use the `ScopeLayoutBuilder`
+    // to encapsulate the logic that can be shared with the entry-point
+    // case.
+    //
+    ScopeLayoutBuilder globalScopeLayoutBuilder;
+    globalScopeLayoutBuilder.beginLayout(&context);
+    for( auto& parameterInfo : sharedContext.parameters )
     {
-        auto globalConstantBufferLayout = createParameterGroupTypeLayout(
-            layoutContext,
-            nullptr,
-            globalScopeRules,
-            globalScopeRules->GetObjectLayout(ShaderParameterKind::ConstantBuffer),
-            globalScopeStructLayout);
-
-        globalScopeLayout = globalConstantBufferLayout;
+        globalScopeLayoutBuilder.addParameter(parameterInfo);
     }
 
-    // We now have a bunch of layout information, which we should
-    // record into a suitable object that represents the program
-    RefPtr<VarLayout> globalVarLayout = new VarLayout();
-    globalVarLayout->typeLayout = globalScopeLayout;
-    if (needDefaultConstantBuffer)
+    auto globalScopeVarLayout = globalScopeLayoutBuilder.endLayout();
+    if( globalConstantBufferBinding.count != 0 )
     {
-        auto cbInfo = globalVarLayout->findOrAddResourceInfo(LayoutResourceKind::ConstantBuffer);
+        auto cbInfo = globalScopeVarLayout->findOrAddResourceInfo(globalConstantBufferBinding.kind);
         cbInfo->space = globalConstantBufferBinding.space;
         cbInfo->index = globalConstantBufferBinding.index;
     }
-    programLayout->globalScopeLayout = globalVarLayout;
+    programLayout->parametersLayout = globalScopeVarLayout;
 }
 
-StructTypeLayout* getGlobalStructLayout(
-    ProgramLayout*  programLayout);
-
 RefPtr<ProgramLayout> specializeProgramLayout(
-    TargetRequest * targetReq,
-    ProgramLayout* programLayout, 
+    TargetRequest*  targetReq,
+    ProgramLayout*  oldProgramLayout, 
     SubstitutionSet typeSubst)
 {
+    // The goal of the layout specialization step is to take an existing `ProgramLayout`,
```

Comment:
I thought at some point we have decided to drop support for global type params. Maybe after this, we just add support for generic entry points and completely remove this special casing logic?

Reply:
Yes, I'm considering this PR as a prerequisite for generic entry points, and I started a branch for generic entry points right after putting up the PR.

The reason for not dropping the existing logic is that we have multiple test cases that currently rely on global generic parameters, and it would be nice to only drop those when we can port them to actually be tests that use generic entry points.

Even with all that said, there is still going to be the problem that users are *probably* going to want us to support global shader parameters of interface types (or types that include them) and those create many of the same challenges as global generics once we start thinking about layout.

Certainly once we drop global generic parameters it would be worth considering whether we can/should change our mind about how layout should be computed in the presence of specialize-able stuff at the global scope. The experiment of trying to keep things as stable as possible was nice in theory, but I'm not actually sure it worked out in practice.